examples/run_on_mdbx.rs

```diff
@@ -0,0 +1,47 @@
+//! Check mainnet blocks using [OnDiskStorage]
+//! For help, run: `cargo run --example run_on_mdbx -- --help`
+
+#![allow(missing_docs)]
+
+use std::{
+    fs::File,
+    io::{BufReader, Error},
+};
+
+use alloy_rpc_types::Block;
+use clap::Parser;
+use pevm::{chain::PevmEthereum, OnDiskStorage, StorageWrapper};
+use revm::db::CacheDB;
+
+#[path = "../tests/common/mod.rs"]
+pub mod common;
+
+/// Check mainnet blocks using [OnDiskStorage]
+#[derive(Parser, Debug)]
+#[clap(name = "run_on_mdbx")]
+struct Args {
+    /// Path to MDBX dir
+    #[clap(long, value_name = "DIR")]
+    mdbx: String,
+    /// Path to block.json file
+    #[clap(long, value_name = "FILE")]
+    block: String,
+}
+
+fn main() -> Result<(), Error> {
+    let args = Args::parse();
+
+    let block: Block = {
+        let file = File::open(args.block)?;
+        serde_json::from_reader(BufReader::new(file))?
+    };
+
+    let on_disk_storage = OnDiskStorage::open(args.mdbx).map_err(Error::other)?;
+    let wrapped_storage = StorageWrapper(&on_disk_storage);
+    let db = CacheDB::new(&wrapped_storage);
+
+    let chain = PevmEthereum::mainnet();
+    common::test_execute_alloy(&db, &chain, block, true);
```

Comment:
Let's port this to `tests`.

Reply:
> Let's port this to an automated test.

Ok, no problem. How about making an automated test that for each block `K`:
1. Read from `data/blocks/<K>/*.json` and write to `data/blocks/<K>/mdbx.dat`
2. Run `pevm` using `OnDiskStorage`
3. Finally, delete `data/blocks/<K>/mdbx.dat`

?

Reply:
Yeah, we can add `OnDiskStorage` to `for_each_block_from_disk`? Then for each block we'll test & bench with both in memory and on disk storage.

Not important now, but eventually we can check in a DB file with all the bytecodes and block hashes as a base DB, to then load and drop accounts for each test/bench.

Reply:
> we can add OnDiskStorage to for_each_block_from_disk

We cannot do this because the function signature is:

```
pub fn for_each_block_from_disk(
  mut handler: impl FnMut(Block, InMemoryStorage)
)
```

instead of:

```
pub fn for_each_block_from_disk(
    mut handler: impl FnMut(Block, Box<dyn Storage<Error = dyn Display>>),
)
```

------

Also, looking at `benches/mainnet.rs`, if we call `handler` twice in `for_each_block_from_disk`, it seems that the second call will override all the output (i.e. `target/criterion/*`) from the first call.

Reply:
My suggestion:

1. Keep `for_each_block_from_disk` as is.
2. Create a new test fn in `tests/mainnet.rs` and a new bench fn in `benches/mainnet.rs`

Reply:
No, I meant for `for_each_block_from_disk` to return both an `InMemoryStorage` and an `OnDiskStorage`.

Reply:
ok, that sounds doable

Reply:
How about this idea?

```rust
struct SampledBlock {
  path_to_block: PathBuf,
  path_to_pre_state: PathBuf,
  path_to_block_hashes: Option<PathBuf>,
}

impl SampledBlock {
  fn get_block(&self) -> Result<Block, std::io::Error>;
  fn get_in_memory_storage(&self, bytecodes: &Bytecodes) -> Result<InMemoryStorage, std::io::Error>;
  fn get_on_disk_storage(&self, bytecodes: &Bytecodes) -> Result<OnDiskStorage, std::io::Error>;
}

fn get_all_sampled_blocks() -> Result<Vec<SampledBlock>, std::io::Error>;
fn get_bytecodes() -> Result<Bytecodes, std::io::Error>;
```

-----

With this struct, the callees can do like:

```rust
let bytecodes = get_bytecodes()?;
for sample in get_all_sampled_blocks()? {
  let block = sample.get_block()?;
  let in_memory_storage = sample.get_in_memory_storage(&bytecodes)?;
  let on_disk_storage = sample.get_on_disk_storage(&bytecodes)?;
  ...
}
```

Reply:
No, we need quick iterative benchmarks to focus on storage optimization. Stop adding code & abstractions that solve (literally) nothing while adding noise to development, code review, and future maintenance costs.

Reply:
ok, understood